File: Segundo/Monopoly/civitas_java (1)/proyecto_civitas/src/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author pc
 */
class JugadorEspeculador extends Jugador  {
    private static int FactorEspeculador=2;
    private int fianza;
    
    JugadorEspeculador(Jugador jugador, int f){
        super(jugador);
        especulador=true;
        fianza=f;
        actualizaPropietarioPorConversion();
    }
    private void actualizaPropietarioPorConversion()
    {
        for(int i=0;i<super.propiedades.size();i++)
        {
            propiedades.get(i).setPropietario(this);
        }
    }
    private boolean puedePagarFianza(){
        boolean puedePagar = false;
        
        if(saldo>=fianza){
            modificarSaldo(-(this.fianza));
            puedePagar = true;
        }
        
        return puedePagar;
    }
    @Override
    protected boolean debeSerEncarcelado(){
        boolean resultado = false;
        if(super.debeSerEncarcelado() ){
            if(!puedePagarFianza())
            resultado = true;
        }
        
        return resultado;
    }
    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        if(encarcelado)
            return false;
   
        if(propiedad.getNumCasas()< getCasasMax())
        return saldo>=propiedad.getPrecioEdificar();
        else
            return false;
    }
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        if(encarcelado)
            return false;
        else
        {
            if(propiedad.getNumHoteles()<getHotelesMax() && propiedad.getNumCasas()>=4)
                return saldo>=propiedad.getPrecioEdificar();
            else
               return false;      
        }
           
    }
    @Override
    boolean pagaImpuesto(float cantidad){
        if(encarcelado)
            return false;
       
        else
            return paga(cantidad/FactorEspeculador);
                    
    }
    @Override
    protected int getCasasMax(){
        return CasasMax*FactorEspeculador;
    }
   
    @Override
    protected int getHotelesMax(){
        return HotelesMax*FactorEspeculador;
    }
    @Override
    public String toString() {
        return "JugadorEspeculador: "+ this.getNombre() + "\n\tfianza : " + fianza +"\n\tencarcelado: " + super.encarcelado  + "\n\tnumCasillaActual=" + this.getNumCasillaActual() + "\n\tpuedeComprar=" + this.getPuedeComprar() + "\n\tsaldo=" + saldo + "\n\tpropiedades=" + propiedades.toString() + "\n\tsalvoconducto=" + super.salvoconducto +"\n";
    }
    @Override
    boolean construirCasa(int ip)
    {
        boolean result=false;
        boolean puedoEdificarCasa=false;

        if(encarcelado)
            return result;
        else{
            boolean existe=this.existeLaPropiedad(ip);
            if(existe)
            {
                TituloPropiedad propiedad=propiedades.get(ip);
                puedoEdificarCasa=this.puedoEdificarCasa(propiedad);
                if(puedoEdificarCasa)
                {
                    result=propiedad.construirCasa(this);
                    if(result)
                        Diario.getInstance().ocurreEvento("El jugador "+this.getNombre()+" construye casa en la propiedad "+propiedades.get(ip).getNombre());
                }

            }
        }
        return result;
       
    }
    @Override
    boolean construirHotel(int ip){
        boolean result=false;
        if(encarcelado)
            return result;

        if(this.existeLaPropiedad(ip))
        {
            TituloPropiedad propiedad=propiedades.get(ip);
            boolean puedoEdificarHotel=this.puedoEdificarHotel(propiedad);
            float precio=propiedad.getPrecioEdificar();
            if(puedoGastar(precio) && propiedad.getNumHoteles()<this.getHotelesMax() && propiedad.getNumCasas()>=this.getCasasPorHotel())
            {
                puedoEdificarHotel=true;
            }
            if(puedoEdificarHotel)
            {
                result=propiedad.construirHotel(this);
                propiedad.derruirCasas(CasasPorHotel, this);
                Diario.getInstance().ocurreEvento("El jugador "+this.getNombre()+" construye hotel en la propiedad "+propiedades.get(ip).getNombre());
            }
        }


       return result; 
    }
    
}

File: Segundo/Monopoly/civitas_ruby (1)/proyecto_civitas_ruby/lib/jugador_especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas

class JugadorEspeculador < Jugador
  @@FactorEspeculador = 2
  def initialize(jugador,fianza)
    jugador_copia(jugador)
    @fianza = fianza
    
    @especulador = true
    
  end
=begin 
  def self.copia(jugador,fianza)
    nuevo_jugador = new(fianza)
    super(jugador)
    puts nuevo_jugador.inspect
    actualiza_propietario_por_conversion
    return nuevo_jugador
  end

  def actualiza_propietario_por_conversion
    for i in 0..@propiedades.size
      @propiedades[i].set_propietario(self)
    end
  end
=end
  
  def debeSerEncarcelado
    resultado = false
    if (super)
      if(!puede_pagar_fianza)
        resultado = true
      end
    end
    return resultado
  end
  
  def puede_pagar_fianza
    puedePagar = false
    if(@saldo >= @fianza)
      modificarSaldo(-@fianza)
      puedePagar = true
    end
    return puedePagar
  end
  
  def puedoEdificarCasa(propiedad)
    if(@encarcelado)
      return false
    end
    if(propiedad.numCasas < (@@CasasMax * @@FactorEspeculador))
      return @saldo >= propiedad.precioEdificar
    else
      return false
    end
  end
  
  def puedoEdificarHotel(propiedad)
    if(@encarcelado)
      return false
    end
    if(propiedad.numHoteles< (@@HotelesMax * @@FactorEspeculador) && propiedad.numCasas >=4)
      return @saldo >= propiedad.precioEdificar
    else
      return false
    end
  end
  
  def pagaImpuesto(cantidad)
    if(@encarcelado)
      return false
    else
      return paga(cantidad/@@FactorEspeculador)
    end
  end
  
  
  
  def getCasasMax
    return (@@CasasMax * @@FactorEspeculador)
  end
  
  def getHotelesMax
    return (@@HotelesMax * @@FactorEspeculador)
  end
  
  def to_string
    a= "JugadorEspeculador{fianza= #{@fianza} , nombre= #{@nombre}, saldo= #{@saldo},encarcelado = #{@encarcelado}, numCasillaActual=  #{@numCasillaActual}, puedeComprar= #{@puedeComprar}, salvoconducto= #{@salvoconducto}, propiedades= "
    for i in 0...@propiedades.size
      a = a + @propiedades[i].nombre
    end
    a = a + "}\n"
    return a
  end
  
  def construirCasa(ip)
    result = false
    puedoEdificarCasa = false
    if(@encarcelado)
      return result
    else
      if (existeLaPropiedad(ip))
        propiedad = @propiedades[ip]
         
        if (puedoEdificarCasa(propiedad))
          result = propiedad.construirCasa(self)
          if (result)
            Diario.instance.ocurre_evento("El jugador #{@nombre} construye una casa en la propiedad "+ @propiedades[ip].nombre)
          end
        end
      end
    end
    return result
  end
  
  def construir_hotel(ip)
    result = false
    if(@encarcelado)
      return result
    end
    
    if(existeLaPropiedad(ip))
      propiedad = @propiedades[ip]
      puedo_EdificarHotel = puedoEdificarHotel(propiedad)
      precio = propiedad.precioEdificar
      if(puedoGastar(precio) && propiedad.numHoteles < @@HotelesMax && propiedad.numCasas >= @@CasasPorHotel)
        puedo_EdificarHotel = true
      end
      if(puedo_EdificarHotel)
        result = propiedad.construirHotel(self)
        propiedad.derruirCasas(@@CasasPorHotel,self)
        Diario.instance.ocurre_evento("El jugador #{@nombre} construye hotel en la propiedad "+ @propiedades[ip].nombre)
      end
    end
    
  end
  
end

end